### PR TITLE
Trac 54596

### PR DESCRIPTION
**Pull request by hellofromtonya**
Fix: block theme with period in name causes blank editor.

Trac ticket: https://core.trac.wordpress.org/ticket/54596

- [x] Apply patch 54596.diff 
- [x] Update the routes in tests
- [x] Add support for `.`, `[`, `]`, `(`, `)`, `_`, and `@`
- [x] Add test data for valid and invalid theme dirnames

## Test Report 
Updated: 14 Dec

Tested with the following theme directory names:

* `wp-content/themes/twentytwentytwo @[0.4.0] ( -_-)` => Works  ✅
* `wp-content/themes/twentytwentytwo-0.4.0` => Works  ✅
* `wp-content/themes/subdir/twentytwentytwo-0.4.0` => Works  ✅
* `wp-content/themes/subdir/twentytwentytwo v0.4.0 @[( -_-)]` => Works  ✅
* `wp-content/themes/subdir/twentytwentytwo {v0.4.0}` => Does not work as `{` and `}` are not valid
* `wp-content/themes/twentytwentytwo {v0.4.0}` => Does not work as `{` and `}` are not valid

Theme in subdirectory: `wp-content/themes/subdir/twentytwentytwo v0.4.0 @[( -_-)]` Works  ✅
test-in-subdir

Theme not in a subdirectory: `wp-content/themes/twentytwentytwo @[0.4.0] ( -_-)/` Works  ✅
test-valid-theme-not-in-subdir

Theme with invalid characters: `wp-content/themes/tt2-{invalid}/` Results in 404 with the following message: "No route was found matching the URL and request method"
test-invalid-theme-dirname

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Review by SergeyBiryukov — APPROVED** [review]

**Comment by spacedmonkey** [comment]
@hellofromtonya I would love to see some unit test with a test data provider, with list of theme names with funny characters, like, {?: etc.

**Comment by hellofromtonya** [comment]
>I would love to see some unit test with a test data provider, with list of theme names with funny characters, like, {?: etc.

Me too, @spacedmonkey. Those data providers are there now.

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 1]
```diff
@@ -132,30 +149,85 @@ public function test_get_theme_item_permission_check() {
...
+		return array(
+			'alphabetic chars'   => array( 'mytheme' ),
+			'alphanumeric chars' => array( 'mythemev1' ),
+			'space'              => array( 'my theme' ),
+			'-'                  => array( 'my-theme' ),
+			'_'                  => array( 'my_theme' ),
+			'.'                  => array( 'mytheme0.1' ),
+			'- and .'            => array( 'my-theme-0.1' ),
+			'space and .'        => array( 'mytheme v0.1' ),
+			'space, -, _, .'     => array( 'my-theme-v0.1' ),
```
Comment (costdev) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php:227]:
```suggestion
			'space, -, _, .'     => array( 'my_theme - v0.1' ),
```
Test data was missing characters mentioned in the dataset's key.

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 2]
```diff
@@ -132,30 +149,85 @@ public function test_get_theme_item_permission_check() {
...
+	 * @ticket 54516
+	 */
+	public function test_get_theme_item( $theme ) {
+		wp_set_current_user( self::$admin_id );
+		switch_theme( $theme );
+
+		$request  = new WP_REST_Request( 'GET', '/wp/v2/global-styles/themes/' . $theme );
 		$response = rest_get_server()->dispatch( $request );
 		$data     = $response->get_data();
 		unset( $data['_links'] );
```
Comment (costdev) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php]: Why do we `unset( $data['_links'] )`?

**Review by costdev — CHANGES_REQUESTED** [review]
PR looks good, just a query for clarification and a correction in a unit test dataset.

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php (hunk 2 above)
Comment (hellofromtonya) [reply tests/phpunit/tests/rest-api/rest-global-styles-controller.php]: No idea. But to minimize changes, I left it for this PR.

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 3]
```diff
@@ -132,30 +149,85 @@ public function test_get_theme_item_permission_check() {
 		$this->assertErrorResponse( 'rest_cannot_manage_global_styles', $response, 403 );
 	}
 
-
 	/**
 	 * @covers WP_REST_Global_Styles_Controller::get_theme_item
 	 * @ticket 54516
```
Comment (hellofromtonya) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php:154]:
```suggestion
	 * @ticket 54516
	 * @ticket 54596
```

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 4]
```diff
@@ -132,30 +149,85 @@ public function test_get_theme_item_permission_check() {
...
-		$request  = new WP_REST_Request( 'GET', '/wp/v2/global-styles/themes/invalid' );
+		$request  = new WP_REST_Request( 'GET', '/wp/v2/global-styles/themes/doesnotexist' );
 		$response = rest_get_server()->dispatch( $request );
 		$this->assertErrorResponse( 'rest_theme_not_found', $response, 404 );
 	}
 
 	/**
+	 * @dataProvider data_get_theme_item_invalid_theme_dirname
 	 * @covers WP_REST_Global_Styles_Controller::get_theme_item
+	 * @ticket 54516
```
Comment (hellofromtonya) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php]:
```suggestion
	 * @ticket 54596
```

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 5]
```diff
@@ -132,30 +149,85 @@ public function test_get_theme_item_permission_check() {
...
+			'with #'                 => array( 'my#theme' ),
+			'with !'                 => array( 'my!theme' ),
+			'multiple invalid chars' => array( 'mytheme-[_(+@)]#! v4.0' ),
+		);
+	}
+
+	/**
+	 * @dataProvider data_get_theme_item
+	 * @covers WP_REST_Global_Styles_Controller::get_theme_item
+	 * @ticket 54516
```
Comment (hellofromtonya) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php]:
```suggestion
	 * @ticket 54596
```

**Review by costdev — APPROVED** [review]

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php [hunk 6]
```diff
@@ -227,10 +227,11 @@ public function get_item_permissions_check( $request ) {
 	 * @return WP_REST_Response|WP_Error
 	 */
 	public function get_item( $request ) {
+		$id = str_replace( '%20', ' ', $request['id'] );
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: This should be moved to `_sanitize_template_id` callback.

Comment (hellofromtonya) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php:230]:
```suggestion
		$id = _sanitize_template_id( $request['id'] );
```

Comment (spacedmonkey) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Also, should we use url decode functions here.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php [hunk 7]
```diff
@@ -522,11 +522,10 @@ public function get_theme_item_permissions_check( $request ) {
 	 * @since 5.9.0
 	 *
 	 * @param WP_REST_Request $request The request instance.
-	 *
 	 * @return WP_REST_Response|WP_Error
 	 */
 	public function get_theme_item( $request ) {
-		if ( wp_get_theme()->get_stylesheet() !== $request['stylesheet'] ) {
+		if ( wp_get_theme()->get_stylesheet() !== str_replace( '%20', ' ', $request['stylesheet'] ) ) {
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php]: We should implement a sanitize callback for stylesheet

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 8]
```diff
@@ -132,30 +149,86 @@ public function test_get_theme_item_permission_check() {
...
 	/**
 	 * @covers WP_REST_Global_Styles_Controller::get_theme_item
 	 * @ticket 54516
+	 * @ticket 54596
 	 */
-	public function test_get_theme_item_invalid() {
+	public function test_get_theme_item_not_found() {
 		wp_set_current_user( self::$admin_id );
-		$request  = new WP_REST_Request( 'GET', '/wp/v2/global-styles/themes/invalid' );
+		$request  = new WP_REST_Request( 'GET', '/wp/v2/global-styles/themes/doesnotexist' );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php]: Why is this changing here?

**Comment by spacedmonkey** [comment]
I would really like @TimothyBJacobs to review this before this is merged.

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php (hunk 8 above)
Comment (hellofromtonya) [reply tests/phpunit/tests/rest-api/rest-global-styles-controller.php]: Carryover from testing theme that does not exist. Reverting.

Comment (hellofromtonya) [reply tests/phpunit/tests/rest-api/rest-global-styles-controller.php]: Reverted in https://github.com/WordPress/wordpress-develop/pull/2034/commits/03bf811e413e6e76d4a864dce0cda339dd930346.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php (hunk 7 above)
Comment (hellofromtonya) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php]: Yup, you're right. Thanks @spacedmonkey. Done in https://github.com/WordPress/wordpress-develop/pull/2034/commits/03bf811e413e6e76d4a864dce0cda339dd930346.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php (hunk 6 above)
Comment (hellofromtonya) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php]: Yup, you're right. Thanks @spacedmonkey. Done in https://github.com/WordPress/wordpress-develop/pull/2034/commits/03bf811e413e6e76d4a864dce0cda339dd930346.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php [hunk 9]
```diff
@@ -522,7 +538,6 @@ public function get_theme_item_permissions_check( $request ) {
 	 * @since 5.9.0
 	 *
 	 * @param WP_REST_Request $request The request instance.
-	 *
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:522]: Why was this line removed

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php [hunk 10]
```diff
@@ -84,6 +86,20 @@ public function register_routes() {
...
 
+	/**
+	 * Sanitize the global styles ID or stylesheet to decode endpoint.
+	 * For example, `wp/v2/global-styles/templatetwentytwo%200.4.0`
+	 * would be decoded to `templatetwentytwo 0.4.0`.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param string $id_or_stylesheet Global styles ID or stylesheet.
+	 * @param string Sanitized global styles ID or stylesheet.
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php]: We could do with a return value here in the comments.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php [hunk 11]
```diff
@@ -41,35 +41,37 @@ public function register_routes() {
...
 					'methods'             => WP_REST_Server::READABLE,
 					'callback'            => array( $this, 'get_item' ),
 					'permission_callback' => array( $this, 'get_item_permissions_check' ),
 					'args'                => array(
 						'id' => array(
-							'description' => __( 'The id of a template' ),
-							'type'        => 'string',
+							'description'       => __( 'The id of a template' ),
+							'type'              => 'string',
+							'sanitize_callback' => array( $this, '_sanitize_global_styles_callback' ),
```
Comment (spacedmonkey) [inline src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:74]: Do we need another method for this? Would this work?
```suggestion
							'sanitize_callback' => 'urldecode',
```

**Comment by spacedmonkey** [comment]
@hellofromtonya This needs to be rebased and conflicts fixed.

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php (hunk 10 above)
Comment (hellofromtonya) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php]:
```suggestion
	 * @return string Sanitized global styles ID or stylesheet.
```

Whoopsie, good catch!

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php (hunk 9 above)
Comment (hellofromtonya) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:522]: Coding standard of no newlines between the params and return. A wee bit of girl scouting.

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 12]
```diff
@@ -145,17 +160,73 @@ public function test_get_theme_item_invalid() {
...
+	 * @return array
+	 */
+	public function data_get_theme_item_invalid_theme_dirname() {
+		return array(
+			'with |'                 => array( 'my|theme' ),
+			'with +'                 => array( 'my+theme' ),
+			'with {}'                => array( 'my{theme}' ),
+			'with #'                 => array( 'my#theme' ),
+			'with !'                 => array( 'my!theme' ),
+			'multiple invalid chars' => array( 'mytheme-[_(+@)]#! v4.0' ),
```
Comment (spacedmonkey) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php:189]: How about some none latin characters, like an è,é,ê,ë. How about a emoji and spaces?

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php [hunk 13]
```diff
@@ -145,17 +160,73 @@ public function test_get_theme_item_invalid() {
...
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_get_theme_item() {
+		return array(
+			'alphabetic chars'   => array( 'mytheme' ),
```
Comment (spacedmonkey) [inline tests/phpunit/tests/rest-api/rest-global-styles-controller.php:219]: https://github.com/WordPress/wordpress-develop/pull/2034/files#r768865447

File: src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php (hunk 11 above)
Comment (hellofromtonya) [reply src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:74]: I did that first but the callback receives 3 args whereas `urldecode()` only accepts 1. Arguments passed are the ID or stylesheet, request object, and the arg name. Tests fail and PHP Warning is thrown on the site.

```
Warning: urldecode() expects exactly 1 parameter, 3 given
```

File: tests/phpunit/tests/rest-api/rest-global-styles-controller.php (hunk 13 above)
Comment (hellofromtonya) [reply tests/phpunit/tests/rest-api/rest-global-styles-controller.php:219]: Testing would need to be done on cross-platforms (i.e. Windows, Mac, Linux, etc) to determine if non-latin characters and/or emojis are valid in filesystem directory names.

**Comment by hellofromtonya** [comment]
Merge conflict resolved.

**Comment by hellofromtonya** [comment]
Committed via https://core.trac.wordpress.org/changeset/52376.
